circuit-foundation/src/commonJvmTest/kotlin/com/slack/circuit/foundation/PausableStateTest.kt

```diff
@@ -0,0 +1,70 @@
+// Copyright (C) 2024 Slack Technologies, LLC
+// SPDX-License-Identifier: Apache-2.0
+package com.slack.circuit.foundation
+
+import androidx.compose.foundation.clickable
+import androidx.compose.foundation.text.BasicText
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.SideEffect
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableIntStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.platform.testTag
+import androidx.compose.ui.test.assertTextEquals
+import androidx.compose.ui.test.junit4.createComposeRule
+import androidx.compose.ui.test.onNodeWithTag
+import androidx.compose.ui.test.performClick
+import com.slack.circuit.runtime.CircuitUiState
+import com.slack.circuit.runtime.presenter.Presenter
+import kotlin.test.assertEquals
+import kotlinx.coroutines.test.runTest
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+
+private const val TEST_TAG = "count"
+private const val EXPECTED_RECOMPOSITIONS = 3
+
+@RunWith(ComposeUiTestRunner::class)
+class PausableStateTest {
+
+  @get:Rule val composeTestRule = createComposeRule()
+
+  @Test
+  fun presentWithLifecycleShouldNotLoop() = runTest {
+    val presenter = TestPresenter()
+    with(composeTestRule) {
+      setContent {
+        val state = presenter.presentWithLifecycle()
+        BasicText(
+          text = "${state.count}",
+          modifier = Modifier.testTag(TEST_TAG).clickable { state.eventSink() },
+        )
+      }
+      with(onNodeWithTag(TEST_TAG)) {
+        assertTextEquals("0")
+        performClick()
+        assertTextEquals("1")
+        performClick()
+        assertTextEquals("2")
+      }
+      assertEquals(EXPECTED_RECOMPOSITIONS, presenter.compositionCount)
+    }
+  }
+
+  class TestPresenter : Presenter<TestState> {
+
+    var compositionCount = 0
+
+    @Composable
+    override fun present(): TestState {
+      var count by remember { mutableIntStateOf(0) }
+      SideEffect { compositionCount++ }
+      return TestState(count = count) { count++ }
+    }
+  }
+
+  data class TestState(val count: Int, val eventSink: () -> Unit) : CircuitUiState
```

Comment:
This state is a data class, so this should pass without any of the other changes in this PR. The issue is for state classes which _don't_ implement equality (which I'd argue is bad design anyway).

The docs on `pausableState` say that the class needs to implement equality for this reason.

Reply:
Ya this is the odd bit, each time there's a `state` write it's triggering a re-compostion which calls `produceState` again. If the second `produceState` call is equal state then it doesn't recompose, which is where non-equal states get stuck looping.

Reply:
Yep, because the `MutableState` can't work out whether the state is the same or not, so triggers it's observation listeners, which results in a recomposition. With an object which implements equality, the `MutableState` will only invoke the listeners if the state being set is different.

When a recomposition happens, it will recompose everything, including the call to `produceState()`, and thus a new state is created and set on the mutable state. Repeat for a composition loop.